assembly/src/parsers/ast/mod.rs

```diff
@@ -0,0 +1,593 @@
+        let full_module_name = self.imports.get(module_name).unwrap();
+        format!("{}{}{}", full_module_name, MODULE_PATH_DELIM, proc_name)
+    }
+
+    // Parse procedures in the source and store them in the program
+    fn parse_procedures(
+        &mut self,
+        tokens: &mut TokenStream,
+        allow_export: bool,
+    ) -> Result<(), AssemblyError> {
+        while let Some(token) = tokens.read() {
+            match token.parts()[0] {
+                Token::EXPORT => {
+                    if !allow_export {
+                        let (label, _, _) = token.parse_proc()?;
+                        return Err(AssemblyError::proc_export_not_allowed(token, &label));
+                    }
+                    self.parse_procedure(tokens)?;
+                }
+                Token::PROC => {
+                    self.parse_procedure(tokens)?;
+                }
+                _ => break,
+            };
+        }
+
+        Ok(())
+    }
+
+    // Parse procedure from token stream and add it to the procedure map in context.
+    fn parse_procedure(&mut self, tokens: &mut TokenStream) -> Result<(), AssemblyError> {
+        let proc_start = tokens.pos();
+
+        // read procedure name and consume the procedure header token
+        let header = tokens.read().expect("missing procedure header");
+        let (label, num_locals, is_export) = header.parse_proc()?;
+        if self.procedures.contains_key(&label) {
+            return Err(AssemblyError::duplicate_proc_label(header, &label));
+        }
+
+        tokens.advance();
+
+        let mut body = Vec::<Node>::new();
+        // parse procedure body
+        self.parse_body(tokens, &mut body)?;
+
+        // consume the 'end' token
+        match tokens.read() {
+            None => Err(AssemblyError::unmatched_proc(
+                tokens.read_at(proc_start).expect("no proc token"),
+            )),
+            Some(token) => match token.parts()[0] {
+                Token::END => token.validate_end(),
+                _ => Err(AssemblyError::unmatched_proc(
+                    tokens.read_at(proc_start).expect("no proc token"),
+                )),
+            },
+        }?;
+        tokens.advance();
+
+        // build and return the procedure
+        let proc = types::Procedure {
+            name: label,
+            num_locals,
+            is_export,
+            body,
+            index: self.procedures.len() as u32,
+        };
+
+        self.procedures.insert(proc.name.to_string(), proc);
+
+        Ok(())
+    }
+
+    // Parses a token from the token stream in a body, which generates a series of AST nodes.
+    fn parse_body(
+        &self,
+        tokens: &mut TokenStream,
+        nodes: &mut Vec<Node>,
+    ) -> Result<(), AssemblyError> {
+        if let Some(token) = tokens.read() {
+            match token.parts()[0] {
+                Token::ELSE => token.validate_else()?,
```

Comment:
Could this actually ever happen? Or should `ELSE` go into unreachable section?

Reply:
hmm I don't think it should happen, it's the same logic we have in the MAST parser so I kept it.
I changed it to return a dangling_else instead